execution_workflows/DaPars2/Dapars2_snakemake/workflow/rules/preprocess_seq.smk

```diff
@@ -0,0 +1,36 @@
+rule generateBedgraph:
+    """This step generates bedgraph files from sample bams. The bedgraph files will then processed by DaPars2.
+    """
+    input:
+        bam=lambda wildcards:
+            pd.Series(samples.loc[wildcards.sample, "bam"]).values
+    output:
+        out=os.path.join(config["out_dir"], "{sample}.bedgraph")
+
+    container:
+        "docker://apaeval/dapars2:1.0"
+    log:
+        os.path.join(LOG_DIR, "{sample}_generateBedgraph.log")
+    shell:
+        """(genomeCoverageBed -bg -ibam {input.bam} >{output.out}) &> {log}"""
+
+
+rule generatePathReadcounts:
+    """This step counts the aligned read of each sample and write the abspaths and the numbers to a txt. The output will be assigned to DaPars2_config_file.txt and used for sequencing depth normalization.
+    """
+    input:
+        bam = lambda wildcards:
+            pd.Series(samples.loc[wildcards.sample, "bam"]).values
+    output:
+        out = os.path.join(config["out_dir"], "{sample}_pathReadcounts.tsv")
+    params:
+        basename = lambda wildcards: os.path.basename(samples.loc[wildcards.sample, "bam"])
+
+    container:
+        "docker://apaeval/dapars2:1.0"
+
+    log:
+        os.path.join(LOG_DIR,"{sample}_generatePathReadcounts.log")
+
+    shell:
+        """(echo "{params.basename}"'\t'$(samtools view -c -F 260 {input.bam}) >{output.out}) &> {log}"""
```

Comment:
I'm a bit uneasy with this as is. Currently, if the samtools command fails the output file is still generated with just 1 column (the sample name) instead of 2 (sample name and read counts). Since the file is generated, Snakemake considers this step complete and the workflow continues, running into errors when DaPars2 is run. I think this could be made more robust by trying something like the following:

```suggestion
    shell:
        """
        counts=$(samtools view -c -F 260 {input.bam})
        echo "{params.basename}"'\t'${counts} > {output.out} 
        &> {log}"""
```

_I think_ by assigning as a variable, it should produce an error rather than giving empty output

Reply:
Resolved with 27a83b3 . Had to double brace `{{counts}}` to prevent Snakemake expanding it. If add dummy error like you suggested the rule will fail and workflow stops here. Workflow runs successfully with test data for this update

```
shell:
    """
    counts=$(samtools view -c -F 260 {input.bam})
    (echo "{params.basename}"'\t'${{counts}} > {output.out}) &> {log}
    """
```